docs/source/tutorials_source/package/tutorial_simsiam_esa.py

```diff
@@ -0,0 +1,447 @@
+        path = os.path.join(path_to_data, filenames[idx])
+        img = Image.open(path)
+        img = functional.resize(img, thumbnail_size)
+        img = np.array(img)
+        img_box = osb.AnnotationBbox(
+            osb.OffsetImage(img, cmap=plt.cm.gray_r),
+            embeddings_2d[idx],
+            pad=0.2,
+        )
+        ax.add_artist(img_box)
+
+    # set aspect ratio
+    ratio = 1. / ax.get_data_ratio()
+    ax.set_aspect(ratio, adjustable='box')
+
+
+# get a scatter plot with thumbnail overlays
+get_scatter_plot_with_thumbnails()
+
+
+# %%
+# Next, we plot example images and their nearest neighbors (calculated from the
+# embeddings generated above). This is a very simple approach to find more images
+# of a certain type where a few examples are already available. For example,
+# when a subset of the data is already labelled and one class of images is clearly
+# underrepresented, one can easily query more images of this class from the 
+# unlabelled dataset.
+#
+# Let's get to work! The plots are shown below.
+
+example_images = [
+    'S2B_MSIL1C_20200526T101559_N0209_R065_T32TNL/tile_00209.png', # water with ships
+    'S2B_MSIL1C_20200526T101559_N0209_R065_T31TGE/tile_00154.png', # water clear 
+    'S2B_MSIL1C_20200526T101559_N0209_R065_T32TMP/tile_00443.png', # coast line
+    'S2B_MSIL1C_20200526T101559_N0209_R065_T32TMM/tile_00040.png', # land 1
+    'S2B_MSIL1C_20200526T101559_N0209_R065_T32TNL/tile_00556.png', # land 2
+    'S2B_MSIL1C_20200526T101559_N0209_R065_T32SLJ/tile_00527.png', # clouds 1
+    'S2B_MSIL1C_20200526T101559_N0209_R065_T31SGD/tile_01731.png', # clouds 2
+    'S2B_MSIL1C_20200526T101559_N0209_R065_T32SMG/tile_00238.png', # clouds 3
+]
+
+
+def get_image_as_np_array(filename: str):
+    """Returns an image as a numpy array.
+    """
+    img = Image.open(filename)
+    return np.asarray(img)
+
+
+def get_image_as_np_array_with_frame(filename: str, w: int = 5):
+    """Returns an image as a numpy array with a black frame of width w.
+    """
+    img = get_image_as_np_array(filename)
+    # get shape of the image
+    ny, nx, _ = img.shape
+    # create an empty image with padding for the frame
+    framed_img = np.zeros((w + ny + w, w + nx + w, 3))
+    framed_img = framed_img.astype(np.uint8)
+    # put the original image in the middle of the new one
+    framed_img[w:-w, w:-w] = img
+    return framed_img
+
+
+def plot_nearest_neighbors_3x3(example_image: str, i: int):
+    """Plots the example image and its eight nearest neighbors.
+    """
+    n_subplots = 9
+    # initialize empty figure
+    fig = plt.figure()
+    fig.suptitle(f"Nearest Neighbor Plot {i + 1}")
+    #
+    example_idx = filenames.index(example_image)
+    # get distances to the cluster center
+    distances = embeddings - embeddings[example_idx]
+    distances = np.power(distances, 2).sum(-1).squeeze()
+    # sort indices by distance to the center
+    argsort = np.argsort(distances)
```

Comment:
'argsort' is quite a bad name, because it is also a function. Furthermore doing a complete sort is not efficient, replace it with `indices_closest_images = np.argpartition(distances,n_subplots-1)[:n_subplots]`

Reply:
The advantage of the complete sort is that the first image is still the example image whereas with argpartition the order of the indices is undefined. Furthermore, this line is by far no bottleneck - I will continue using the full sort but change the name.

Reply:
If it is not the bottleneck, then the performance is indeed irrelevant. I didn't think about the order within the 9 nearest images being relevant, sorry.